Distance_3/include/CGAL/squared_distance_3_1.h

```diff
@@ -44,6 +44,22 @@ squared_distance(
   return internal::squared_distance_to_line(dir, diff, k);
 }
 
+template <class K>
+void
+squared_distance_RT(
+    const typename K::Point_3 &pt,
+    const typename K::Line_3 &line,
+    typename K::RT& num,
+    typename K::RT& den,
+    const K& k)
+{
+  typedef typename K::Vector_3 Vector_3;
+  typename K::Construct_vector_3 construct_vector;
+  Vector_3 dir(line.direction().vector());
+  Vector_3 diff = construct_vector(line.point(), pt);
```

Comment:
That's completely ignoring kernel concepts, but the whole package is afflicted... Might do it in the other PR if I have time.

Reply:
I think that is fine: the bodies of functions of the kernel are implementation details, and they are free to use whatever API they want.

Reply:
If you look at it as the implementation of `Compute_squared_distance_3`, sure; but the global functions go straight for `internal::squared_distance(O1, O2, K())` and `K` is not necessarily a CGAL kernel.

I guess the error is that the global function should actually call `K().compute_squared_distance_3()(O1, O2)`, which THEN will call `internal::squared_distance(O1, O2)` in the case of CGAL kernels, and in the case of (very hypothetical) third party kernels it will go to their implementation.

Reply:
> If you look at it as the implementation of `Compute_squared_distance_3`, sure; but the global functions go straight for `internal::squared_distance(O1, O2, K())` and `K` is not necessarily a CGAL kernel.

As far as I know, global functions are not supposed to work for anything but CGAL kernels: in `CGAL::Point_3<K>`, the `K` must be one of the CGAL kernels. Maybe we want to change that fact, but for the moment that is hard-coded in a lot of places in the CGAL kernel code.

Reply:
Right, I forgot that only the internal global function uses Kernel::Object, so it's indeed only CGAL kernels.